### PR TITLE
Task 5

### PR DESCRIPTION
**Pull request by klimashevsky-pavel**
task - https://epam-my.sharepoint.com/personal/vitali_kozlov_epam_com/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fvitali_kozlov_epam_com%2FDocuments%2FNode%2Ejs%2FCDP%2FHomeworks%2F5%2E%20Authorization%2E%20Validation%2E%20Security%2FHomework%205%2Epdf&parent=%2Fpersonal%2Fvitali_kozlov_epam_com%2FDocuments%2FNode%2Ejs%2FCDP%2FHomeworks%2F5%2E%20Authorization%2E%20Validation%2E%20Security
Implemented Facebook strategy, skipped Twitter and Google.

### CODE REVIEW

File: src/config/users.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        password: 'I_am_DNO'
+    },
+    {
+        userName: 'Pavel Klimasheuski',
+        password: 'I_am_Pasha'
+    },
+    {
+        userName: 'Palina Kartsel',
+        password: 'I_am_Polina'
+    }
```
Comment (deser) [inline src/config/users.ts:18]: dad, mom, grandfather, etc?

File: src/controllers/authController.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            message: 'Authentication failed! Please check the request'
+        });
+    }
+
+    const tokenData = formTokenDataObject(userName);
+    const response = {
+        code: 200,
+        message: 'OK',
+        data: {
+            user: userFromDB
```
Comment (deser) [inline src/controllers/authController.ts:43]: is it ok to expose user pwd and personal info? :)

File: src/controllers/authController.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        userName
+    });
+    refreshTokensCollection.push({ userName, refreshToken: tokenData.refreshToken });
+    res.status(200).json({ ...response });
+};
+
+const verifyToken = (req: ExtendedRequest, res: Response, next: () => void) => {
+    const authHeader = req.headers[AUTHORIZATION_HEADER_KEY];
+    const token =
+        authHeader && authHeader.startsWith('Bearer ')
```
Comment (deser) [inline src/controllers/authController.ts:82]: lodash

File: src/controllers/authController.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            ? authHeader.slice(TOKEN_START_INDEX)
+            : authHeader;
+
+    if (token) {
+        jwt.verify(token, config.jwt_secret, err => {
+            if (err) {
+                return res.json({
+                    success: false,
```
Comment (deser) [inline src/controllers/authController.ts:90]: any http codes?

File: src/controllers/authController.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                    message: 'Token is not valid'
+                });
+            } else {
+                next();
+            }
+        });
+    } else {
+        return res.json({
+            success: false,
```
Comment (deser) [inline src/controllers/authController.ts:99]: any http codes?

File: src/controllers/authController.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    refreshTokensCollection = reject<RefreshToken>(refreshTokensCollection, {
+        refreshToken,
+        userName
+    });
+    refreshTokensCollection.push({ userName, refreshToken: tokenData.refreshToken });
+    res.status(200).json({ ...response });
+};
+
+const verifyToken = (req: ExtendedRequest, res: Response, next: () => void) => {
```
Comment (deser) [inline src/controllers/authController.ts:79]: Seems verifyToken should know nothing about request\response. It's function is to take a string and say if it is valid token or not

File: src/http-servers/plain-text-server.ts [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import http, { ServerResponse, IncomingMessage } from 'http';
 
-const PORT = 3001;
+const PORT = 3005;
```
Comment (deser) [inline src/http-servers/plain-text-server.ts:3]: to config? :)

File: src/middlewares/cookieParser.ts [hunk 8]
```diff
@@ -8,6 +8,7 @@ const cookieParser = (req: ExtendedRequest, res: Response, next: () => void) =>
 };
 
 const parseCookies = (cookieString: string): object => {
+    if (!cookieString) return {};
     const cookiesArray = cookieString.split('; ');
```
Comment (deser) [inline src/middlewares/cookieParser.ts:12]: why not lodash?

File: src/utils/addRoutes.ts [hunk 9]
```diff
@@ -1,13 +1,40 @@
...
+        '/api/products/:id',
+        authController.verifyToken,
+        productsController.getSingleProduct
+    );
+    router.get(
+        '/api/products/:id/reviews',
+        authController.verifyToken,
+        productsController.getSingleProductReviews
+    );
+    router.post('/api/products', authController.verifyToken, productsController.addProduct);
```
Comment (deser) [inline src/utils/addRoutes.ts:36]: hm, adding verifyToken to every request is not correct, it should be somewhere in middlewares

File: src/utils/setupPassport.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+    passport.use(
+        new FacebookStrategy(
+            {
+                clientID: '2237257873251788',
+                clientSecret: 'a5cc1fd744f81f2c0ae81e835799007d',
+                callbackURL: 'http://localhost:8080/login/facebook/callback'
+            },
+            function(accessToken, refreshToken, profile, cb) {
+                console.log('profile----------->', profile);
```
Comment (deser) [inline src/utils/setupPassport.ts:28]: remove

File: src/config/users.ts (hunk 1 above)
Comment (sleeping-cucumber) [reply src/config/users.ts:18]: and dog as well!

File: src/controllers/authController.ts (hunk 6 above)
Comment (sleeping-cucumber) [reply src/controllers/authController.ts:79]: It's a name of a controller, it's middleware in fact. May be should be renamed (but I think current name is ok), but workflow is right I guess

**Pull request merged**